### PR TITLE
[test-integration] Refactor and extract TestRequire functionality

### PR DESCRIPTION
**Pull request by vdemeester**
This will help when extracting suites in their own package. Ideas is :

```go
func (s *MySuite) TestMyTest(c *check.C) {
    requirement.Is(c, SameDaemon, ArchitectureIs("amd64"), someLocalCheck)
    // […]
}
```

I really wish `go-check` would use `Skip(...interface{})` instead of `Skip(string)`, as it would make this mechanism work for both `go-check` and `testing` packages.

/cc @icecrime @thaJeztah @dnephin @cpuguy83 

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by thaJeztah** [comment]
SGTM

File: integration-cli/requirement/requirement.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// Test represent a function that can be used as a requirement validation.
+type Test func() (bool, string)
+
+// Is checks if the environment satisfies the requirements
+// for the test to run or skips the tests.
+func Is(s skipT, requirements ...Test) {
+	for _, r := range requirements {
+		isValid, message := r()
+		if !isValid {
+			s.Skip(message)
```
Comment (icecrime) [inline integration-cli/requirement/requirement.go]: If that's the only place where the message is used, then _maybe_ we should consider a nuclear option that would avoid complicating the syntax for all the callers of requirement functions. We could:

- Have requirements functions be `func () bool`, in such way that we can continue writing `if SameHostDaemon() {`.
- Use some reflect magic here to provide a message, as in `s.Skip(fmt.Sprintf("unmatched requirement %s", runtime.FuncForPC(<some_reflect_magic>).Name())`.

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by tiborvass** [comment]
@vdemeester needs a rebase

**Comment by thaJeztah** [comment]
@vdemeester is rebasing :house: -> 🏡

**Comment by vdemeester** [comment]
🏡  and rebased

File: integration-cli/requirements_unix_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+func overlaySupported() bool {
+	cmd := exec.Command(dockerBinary, "run", "--rm", "busybox", "/bin/sh", "-c", "cat /proc/filesystems")
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return false
+	}
+	return bytes.Contains(out, []byte("overlay\n"))
+}
+
+func overlayFSSupported() bool {
```
Comment (cpuguy83) [inline integration-cli/requirements_unix_test.go]: Should just be `overlayFSSupported` and remove the `overlaySupported`

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: integration-cli/requirements_unix_test.go (hunk 2 above)
Comment (vdemeester) [inline integration-cli/requirements_unix_test.go]: @cpuguy83 oh thanks… I messed up the *conflict* merging here…

**Comment by vdemeester** [comment]
@cpuguy83 done :angel:

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**